### PR TITLE
Maxpool. Kernel of size 1: stride being ignored

### PR DESCRIPTION
**Pull request by lakhinderwalia**
A Resnet50 model isn't being verified properly. The error stems from a JIT maxpool kernel.

### CODE REVIEW

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp [hunk 1]
```diff
@@ -133,9 +133,8 @@ struct window
         auto win_start = generate_array<diff_int>(rank{}, [&](auto j) {
             diff_int w   = win[j];
             diff_int dim = i[j];
+	    (void) w;
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:136]: FYI, testing of JIT pooling in CI disabled. 

https://github.com/ROCm/AMDMIGraphX/issues/3234

Comment (lakhinderwalia) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:136]: Thanks. This is already being verified by forcing it manually.

Comment (umangyadav) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp]: > This is already being verified by forcing it manually -- in a separate build. 
Where ?

File: src/include/migraphx/op/pooling.hpp [hunk 2]
```diff
@@ -331,7 +331,7 @@ struct pooling
                 int end;
                 std::size_t dilated_kernel_dim = dilate_dim(kernel_dims[d_2], dilations[d_2]);
                 // NOLINT
-                if(count_include_pad and (mode != pooling_mode::max))
+                if(count_include_pad and mode == pooling_mode::average)
```
Comment (umangyadav) [inline src/include/migraphx/op/pooling.hpp:334]: This would disable count_include_pad for the lp_norm

Comment (lakhinderwalia) [reply src/include/migraphx/op/pooling.hpp:334]: https://onnx.ai/onnx/operators/onnx__LpNormalization.html

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp (hunk 1 above)
Comment (lakhinderwalia) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp]: Sorry, I meant in a private build. This is being tested by disabling this clause: https://github.com/ROCm/AMDMIGraphX/blob/b8c33b0754041cd61a0bd41396a1838b7c88a8c7/src/targets/gpu/lowering.cpp#L314, to turn on JIT kernels for pooling. So this will test all the pooling modes. `make check` and `test_verify` have been run successfully.

File: test/verify/test_max_pooling_1x1_s2.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+struct test_max_pooling_1x1_s2 : verify_program<test_max_pooling_1x1_s2>
+{
+    migraphx::program create_program() const
+    {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+
+        auto input =
+            mm->add_parameter("x", migraphx::shape{migraphx::shape::float_type, {1, 4, 112, 112}});
```
Comment (umangyadav) [inline test/verify/test_max_pooling_1x1_s2.cpp:39]: You can probably make dimensions smaller for the unit-test to not slow down verify test.

**Review by umangyadav — APPROVED** [review]

File: test/verify/test_max_pooling_1x1_s2.cpp (hunk 3 above)
Comment (lakhinderwalia) [reply test/verify/test_max_pooling_1x1_s2.cpp]: Done.

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp [hunk 4]
```diff
@@ -133,9 +133,8 @@ struct window
         auto win_start = generate_array<diff_int>(rank{}, [&](auto j) {
             diff_int w   = win[j];
             diff_int dim = i[j];
+            (void)w;
             MIGRAPHX_ASSERT(w >= 1);
-            if(w == 1)
-                return dim;
```
Comment (pfultz2) [inline src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:138]: Wouldn't it be better to change this to `return dim * s` instead of removing the condition?

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp [hunk 5]
```diff
@@ -133,9 +133,8 @@ struct window
         auto win_start = generate_array<diff_int>(rank{}, [&](auto j) {
             diff_int w   = win[j];
             diff_int dim = i[j];
+            (void)w;
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:136]: instead of adding `(void)(w)` it might be better to do 
`MIGRAPHX_ASSERT(win[j] >= 1)`

Comment (lakhinderwalia) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:136]: In some build, where assert is not active, the void statement is necessary-- for compilation. Adding another assert won't help in that direction.

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp (hunk 4 above)
Comment (lakhinderwalia) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:138]: If the rest of the following lines do the same thing.. why add an exception?

**Comment by causten** [comment]
Would like to move to 6.2.1 as well

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp (hunk 5 above)
Comment (pfultz2) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:136]: Umang is saying to remove the variable instead of using a void statement.

File: src/include/migraphx/op/pooling.hpp (hunk 2 above)
Comment (pfultz2) [inline src/include/migraphx/op/pooling.hpp:334]: Why is this change needed? The `count_include_pad` flag should only be skipped for max since it doesn't use an average of the pixel values.

File: src/targets/gpu/lowering.cpp [hunk 6]
```diff
@@ -303,7 +303,8 @@ struct miopen_apply
         apply_map.emplace("pooling", [=](instruction_ref ins) {
             auto&& op   = ins->get_operator();
             auto op_val = op.to_value();
-            if(op_val.at("count_include_pad").to<bool>())
+            if(op_val.at("count_include_pad").to<bool>() and
+               op_val["mode"].to<op::pooling_mode>() == op::pooling_mode::average)
```
Comment (pfultz2) [inline src/targets/gpu/lowering.cpp:307]:
```suggestion
            if(op_val.at("count_include_pad").to<bool>() and
               op_val["mode"].to<op::pooling_mode>() != op::pooling_mode::max)
```

File: src/include/migraphx/op/pooling.hpp (hunk 2 above)
Comment (lakhinderwalia) [reply src/include/migraphx/op/pooling.hpp:334]: Per Onnx specfications, count_include_pad as an operator parameter is used only with pooling_mode::average. Not with other pooling modes, e.g. max or lpnorm.

Comment (pfultz2) [reply src/include/migraphx/op/pooling.hpp:334]: This is not an onnx operator. We have this attribute on this operator and it should work for all pooling modes that use a count.

Comment (pfultz2) [reply src/include/migraphx/op/pooling.hpp:334]: Ah wait, lpnorm doesnt use the count. Nevermind.

**Review by bpickrel — CHANGES_REQUESTED** [review]
see earlier comment about the `count_include_pad` param

File: src/include/migraphx/op/pooling.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pfultz2) [reply src/include/migraphx/op/pooling.hpp:334]: lpnorm and max pooling doesn't use a count so this flag is not applicable. You would get the same result if it were true or false.

**Review by pfultz2 — APPROVED** [review]

File: src/include/migraphx/op/pooling.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lakhinderwalia) [reply src/include/migraphx/op/pooling.hpp:334]: (Thanks, @pfultz2 for responding to this. I had missed this comment from Brian.)

**Review by bpickrel — APPROVED** [review]
Assuming the tests pass, should be good.
